diff -urN ati-gatos-4.4.0.orig/r128.h ati-gatos-4.4.0/r128.h
--- ati-gatos-4.4.0.orig/r128.h	2004-08-20 20:41:46.361881992 -0400
+++ ati-gatos-4.4.0/r128.h	2004-08-20 20:42:19.289876176 -0400
@@ -289,19 +289,19 @@
     Bool              directRenderingEnabled;
     DRIInfoPtr        pDRIInfo;
     int               drmFD;
-    drmContext        drmCtx;
+    drm_context_t        drmCtx;
     int               numVisualConfigs;
     __GLXvisualConfig *pVisualConfigs;
     R128ConfigPrivPtr pVisualConfigsPriv;
 
-    drmHandle         fbHandle;
+    drm_handle_t         fbHandle;
 
     drmSize           registerSize;
-    drmHandle         registerHandle;
+    drm_handle_t         registerHandle;
 
     Bool              IsPCI;            /* Current card is a PCI card */
     drmSize           pciSize;
-    drmHandle         pciMemHandle;
+    drm_handle_t         pciMemHandle;
     unsigned char     *PCI;             /* Map */
 
     Bool              allowPageFlip;    /* Enable 3d page flipping */
@@ -309,7 +309,7 @@
     int               drmMinor;
 
     drmSize           agpSize;
-    drmHandle         agpMemHandle;     /* Handle from drmAgpAlloc */
+    drm_handle_t         agpMemHandle;     /* Handle from drmAgpAlloc */
     unsigned long     agpOffset;
     unsigned char     *AGP;             /* Map */
     int               agpMode;
@@ -322,20 +322,20 @@
 
 				/* CCE ring buffer data */
     unsigned long     ringStart;        /* Offset into AGP space */
-    drmHandle         ringHandle;       /* Handle from drmAddMap */
+    drm_handle_t         ringHandle;       /* Handle from drmAddMap */
     drmSize           ringMapSize;      /* Size of map */
     int               ringSize;         /* Size of ring (in MB) */
     unsigned char     *ring;            /* Map */
     int               ringSizeLog2QW;
 
     unsigned long     ringReadOffset;   /* Offset into AGP space */
-    drmHandle         ringReadPtrHandle; /* Handle from drmAddMap */
+    drm_handle_t         ringReadPtrHandle; /* Handle from drmAddMap */
     drmSize           ringReadMapSize;  /* Size of map */
     unsigned char     *ringReadPtr;     /* Map */
 
 				/* CCE vertex/indirect buffer data */
     unsigned long     bufStart;        /* Offset into AGP space */
-    drmHandle         bufHandle;       /* Handle from drmAddMap */
+    drm_handle_t         bufHandle;       /* Handle from drmAddMap */
     drmSize           bufMapSize;      /* Size of map */
     int               bufSize;         /* Size of buffers (in MB) */
     unsigned char     *buf;            /* Map */
@@ -344,7 +344,7 @@
 
 				/* CCE AGP Texture data */
     unsigned long     agpTexStart;      /* Offset into AGP space */
-    drmHandle         agpTexHandle;     /* Handle from drmAddMap */
+    drm_handle_t         agpTexHandle;     /* Handle from drmAddMap */
     drmSize           agpTexMapSize;    /* Size of map */
     int               agpTexSize;       /* Size of AGP tex space (in MB) */
     unsigned char     *agpTex;          /* Map */
diff -urN ati-gatos-4.4.0.orig/r128_dri.c ati-gatos-4.4.0/r128_dri.c
--- ati-gatos-4.4.0.orig/r128_dri.c	2004-08-20 20:41:46.363881688 -0400
+++ ati-gatos-4.4.0/r128_dri.c	2004-08-20 20:42:19.313872528 -0400
@@ -277,7 +277,7 @@
 
 /* Create the Rage 128-specific context information */
 static Bool R128CreateContext(ScreenPtr pScreen, VisualPtr visual,
-			      drmContext hwContext, void *pVisualConfigPriv,
+			      drm_context_t hwContext, void *pVisualConfigPriv,
 			      DRIContextType contextStore)
 {
     ScrnInfoPtr pScrn = xf86Screens[pScreen->myNum];
@@ -288,7 +288,7 @@
 }
 
 /* Destroy the Rage 128-specific context information */
-static void R128DestroyContext(ScreenPtr pScreen, drmContext hwContext,
+static void R128DestroyContext(ScreenPtr pScreen, drm_context_t hwContext,
 			       DRIContextType contextStore)
 {
     /* Nothing yet */
diff -urN ati-gatos-4.4.0.orig/r128_dri.h ati-gatos-4.4.0/r128_dri.h
--- ati-gatos-4.4.0.orig/r128_dri.h	2004-08-20 20:41:46.363881688 -0400
+++ ati-gatos-4.4.0/r128_dri.h	2004-08-20 20:42:19.314872376 -0400
@@ -88,11 +88,11 @@
     int           log2TexGran;
 
     /* MMIO register data */
-    drmHandle     registerHandle;
+    drm_handle_t     registerHandle;
     drmSize       registerSize;
 
     /* CCE AGP Texture data */
-    drmHandle     agpTexHandle;
+    drm_handle_t     agpTexHandle;
     drmSize       agpTexMapSize;
     int           log2AGPTexGran;
     int           agpTexOffset;
diff -urN ati-gatos-4.4.0.orig/r128_sarea.h ati-gatos-4.4.0/r128_sarea.h
--- ati-gatos-4.4.0.orig/r128_sarea.h	2004-08-20 20:41:46.364881536 -0400
+++ ati-gatos-4.4.0/r128_sarea.h	2004-08-20 20:42:19.349867056 -0400
@@ -167,7 +167,7 @@
 #ifdef XF86DRI
     /* The current cliprects, or a subset thereof.
      */
-    XF86DRIClipRectRec boxes[R128_NR_SAREA_CLIPRECTS];
+    drm_clip_rect_t boxes[R128_NR_SAREA_CLIPRECTS];
     unsigned int nbox;
 #endif
 
diff -urN ati-gatos-4.4.0.orig/radeon.h ati-gatos-4.4.0/radeon.h
--- ati-gatos-4.4.0.orig/radeon.h	2004-08-20 20:41:46.365881384 -0400
+++ ati-gatos-4.4.0/radeon.h	2004-08-20 20:42:19.370863864 -0400
@@ -412,14 +412,14 @@
     __GLXvisualConfig *pVisualConfigs;
     RADEONConfigPrivPtr pVisualConfigsPriv;
 
-    drmHandle         fbHandle;
+    drm_handle_t         fbHandle;
 
     drmSize           registerSize;
-    drmHandle         registerHandle;
+    drm_handle_t         registerHandle;
 
     Bool              IsPCI;            /* Current card is a PCI card */
     drmSize           pciSize;
-    drmHandle         pciMemHandle;
+    drm_handle_t         pciMemHandle;
     unsigned char     *PCI;             /* Map */
 
     Bool              depthMoves;       /* Enable depth moves -- slow! */
@@ -428,7 +428,7 @@
     int               drmMinor;
 
     drmSize           gartSize;
-    drmHandle         agpMemHandle;     /* Handle from drmAgpAlloc */
+    drm_handle_t         agpMemHandle;     /* Handle from drmAgpAlloc */
     unsigned long     gartOffset;
     unsigned char     *AGP;             /* Map */
     int               agpMode;
@@ -445,20 +445,20 @@
 
 				/* CP ring buffer data */
     unsigned long     ringStart;        /* Offset into GART space */
-    drmHandle         ringHandle;       /* Handle from drmAddMap */
+    drm_handle_t         ringHandle;       /* Handle from drmAddMap */
     drmSize           ringMapSize;      /* Size of map */
     int               ringSize;         /* Size of ring (in MB) */
     unsigned char     *ring;            /* Map */
     int               ringSizeLog2QW;
 
     unsigned long     ringReadOffset;   /* Offset into GART space */
-    drmHandle         ringReadPtrHandle; /* Handle from drmAddMap */
+    drm_handle_t         ringReadPtrHandle; /* Handle from drmAddMap */
     drmSize           ringReadMapSize;  /* Size of map */
     unsigned char     *ringReadPtr;     /* Map */
 
 				/* CP vertex/indirect buffer data */
     unsigned long     bufStart;         /* Offset into GART space */
-    drmHandle         bufHandle;        /* Handle from drmAddMap */
+    drm_handle_t         bufHandle;        /* Handle from drmAddMap */
     drmSize           bufMapSize;       /* Size of map */
     int               bufSize;          /* Size of buffers (in MB) */
     unsigned char     *buf;             /* Map */
@@ -467,7 +467,7 @@
 
 				/* CP GART Texture data */
     unsigned long     gartTexStart;      /* Offset into GART space */
-    drmHandle         gartTexHandle;     /* Handle from drmAddMap */
+    drm_handle_t         gartTexHandle;     /* Handle from drmAddMap */
     drmSize           gartTexMapSize;    /* Size of map */
     int               gartTexSize;       /* Size of GART tex space (in MB) */
     unsigned char     *gartTex;          /* Map */
diff -urN ati-gatos-4.4.0.orig/radeon_common.h ati-gatos-4.4.0/radeon_common.h
--- ati-gatos-4.4.0.orig/radeon_common.h	2004-08-20 20:41:46.366881232 -0400
+++ ati-gatos-4.4.0/radeon_common.h	2004-08-20 20:42:19.384861736 -0400
@@ -276,7 +276,7 @@
 	int bufsz;
 	char *buf;
 	int nbox;
-        drmClipRect *boxes;
+        drm_clip_rect_t *boxes;
 } drmRadeonCmdBuffer;
 
 /* New style per-packet identifiers for use in cmd_buffer ioctl with
diff -urN ati-gatos-4.4.0.orig/radeon_dri.c ati-gatos-4.4.0/radeon_dri.c
--- ati-gatos-4.4.0.orig/radeon_dri.c	2004-08-20 20:41:46.367881080 -0400
+++ ati-gatos-4.4.0/radeon_dri.c	2004-08-20 20:42:19.400859304 -0400
@@ -284,7 +284,7 @@
 
 /* Create the Radeon-specific context information */
 static Bool RADEONCreateContext(ScreenPtr pScreen, VisualPtr visual,
-				drmContext hwContext, void *pVisualConfigPriv,
+				drm_context_t hwContext, void *pVisualConfigPriv,
 				DRIContextType contextStore)
 {
 #ifdef PER_CONTEXT_SAREA
@@ -321,7 +321,7 @@
 }
 
 /* Destroy the Radeon-specific context information */
-static void RADEONDestroyContext(ScreenPtr pScreen, drmContext hwContext,
+static void RADEONDestroyContext(ScreenPtr pScreen, drm_context_t hwContext,
 				 DRIContextType contextStore)
 {
 #ifdef PER_CONTEXT_SAREA
diff -urN ati-gatos-4.4.0.orig/radeon_dri.h ati-gatos-4.4.0/radeon_dri.h
--- ati-gatos-4.4.0.orig/radeon_dri.h	2004-08-20 20:41:46.367881080 -0400
+++ ati-gatos-4.4.0/radeon_dri.h	2004-08-20 20:42:19.401859152 -0400
@@ -85,15 +85,15 @@
     int           log2TexGran;
 
     /* MMIO register data */
-    drmHandle     registerHandle;
+    drm_handle_t     registerHandle;
     drmSize       registerSize;
 
     /* CP in-memory status information */
-    drmHandle     statusHandle;
+    drm_handle_t     statusHandle;
     drmSize       statusSize;
 
     /* CP GART Texture data */
-    drmHandle     gartTexHandle;
+    drm_handle_t     gartTexHandle;
     drmSize       gartTexMapSize;
     int           log2GARTTexGran;
     int           gartTexOffset;
diff -urN ati-gatos-4.4.0.orig/radeon_dripriv.h ati-gatos-4.4.0/radeon_dripriv.h
--- ati-gatos-4.4.0.orig/radeon_dripriv.h	2004-08-20 20:41:46.367881080 -0400
+++ ati-gatos-4.4.0/radeon_dripriv.h	2004-08-20 20:42:19.402859000 -0400
@@ -53,8 +53,8 @@
 
 typedef struct {
 #ifdef PER_CONTEXT_SAREA
-    drmContext ctx_id;
-    drmHandle sarea_handle;
+    drm_context_t ctx_id;
+    drm_handle_t sarea_handle;
 #else
     /* Nothing here yet */
     int dummy;
diff -urN ati-gatos-4.4.0.orig/radeon_sarea.h ati-gatos-4.4.0/radeon_sarea.h
--- ati-gatos-4.4.0.orig/radeon_sarea.h	2004-08-20 20:41:46.370880624 -0400
+++ ati-gatos-4.4.0/radeon_sarea.h	2004-08-20 20:42:19.461850032 -0400
@@ -201,7 +201,7 @@
     unsigned int vc_format;
 
     /* The current cliprects, or a subset thereof */
-    XF86DRIClipRectRec boxes[RADEON_NR_SAREA_CLIPRECTS];
+    drm_clip_rect_t boxes[RADEON_NR_SAREA_CLIPRECTS];
     unsigned int nbox;
 
     /* Counters for throttling of rendering clients */
