diff -urNp ati-driver-8.28.8.old/common/lib/modules/fglrx/build_mod/firegl_public.c ati-driver-8.28.8/common/lib/modules/fglrx/build_mod/firegl_public.c
--- ati-driver-8.28.8.old/common/lib/modules/fglrx/build_mod/firegl_public.c	2006-08-17 19:10:13.000000000 +0300
+++ ati-driver-8.28.8/common/lib/modules/fglrx/build_mod/firegl_public.c	2006-09-21 10:47:02.000000000 +0300
@@ -23,6 +23,11 @@
 
 // ============================================================
 #include <linux/version.h>
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,18)
+/* No longer valid */
+#define VM_SHM 0
+#include <linux/utsrelease.h>
+#endif
 #ifdef MODVERSIONS
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,5,71)
 #include <linux/modversions.h>
