--- picard-0.10/picard/musicdns/avcodec.c.org   2008-11-09 12:47:41.000000000 +0100
+++ picard-0.10/picard/musicdns/avcodec.c       2008-11-09 12:47:49.000000000 +0100
@@ -278,7 +278,8 @@
         data = packet.data;

         while (size > 0) {
-            len = avcodec_decode_audio(codec_context, (int16_t *)buffer_ptr, &output_size, data, size);
+            output_size = buffer_size + AVCODEC_MAX_AUDIO_FRAME_SIZE;
+            len = avcodec_decode_audio2(codec_context, (int16_t *)buffer_ptr, &output_size, data, size);

             if (len < 0)
                 break;
