--- command.c.orig	2007-02-24 23:36:05.000000000 +0100
+++ command.c	2007-02-24 23:40:49.000000000 +0100
@@ -1817,6 +1817,10 @@
 		    osd_level = v > max ? max : v;
 		/* Show OSD state when disabled, but not when an explicit
 		   argument is given to the OSD command, i.e. in slave mode. */
+		if (v == -1) {
+			if (osd_level == 0) osd_level = 1;
+			if (osd_level == 2) osd_level = 3;
+		}
 		if (v == -1 && osd_level <= 1)
 		    set_osd_msg(OSD_MSG_OSD_STATUS, 0, osd_duration,
 				MSGTR_OSDosd,
