--- mplayer-orig/Makefile	2008-05-28 14:28:39.000000000 +0200
+++ mplayer/Makefile	2008-05-28 14:29:32.000000000 +0200
@@ -745,12 +745,6 @@
 help_mp.h: help/help_mp-en.h $(HELP_FILE)
 	help/help_create.sh $(HELP_FILE) $(CHARSET)
 
-# rebuild version.h each time the working copy is updated
-ifeq ($(wildcard .svn/entries),.svn/entries)
-version.h: .svn/entries
-endif
-version.h:
-	./version.sh `$(CC) -dumpversion`
 
 osdep/mplayer-rc.o: osdep/mplayer.rc version.h
 	$(WINDRES) -I. -o $@ $<
