Index: libswscale/yuv2rgb.c
===================================================================
--- libswscale/yuv2rgb.c        (Revision 27482)
+++ libswscale/yuv2rgb.c        (Arbeitskopie)
@@ -157,6 +157,9 @@

 // The volatile is required because gcc otherwise optimizes some writes away
 // not knowing that these are read in the ASM block.
+#ifdef __ICC
+#define static
+#endif
 static volatile uint64_t attribute_used __attribute__((aligned(8))) b5Dither;
 static volatile uint64_t attribute_used __attribute__((aligned(8))) g5Dither;
 static volatile uint64_t attribute_used __attribute__((aligned(8))) g6Dither;
