--- flock.orig/layout/build/Makefile.in	2007-11-03 06:50:41.000000000 +0900
+++ flock/layout/build/Makefile.in	2007-11-06 01:18:39.000000000 +0900
@@ -212,9 +212,15 @@
 ifdef MOZ_ENABLE_XLIB
 EXTRA_DSO_LDOPTS += \
 		-lxlibrgb \
+		$(XLDFLAGS) $(XT_LIBS) $(XLIBS) \
 		$(NULL)
 endif
 
+ifndef MOZ_ENABLE_LIBXFT
+EXTRA_DSO_LDOPTS += $(MOZ_XFT_LIBS) \
+		    $(NULL)
+endif
+
 ifdef NS_TRACE_MALLOC
 EXTRA_DSO_LIBS	+= tracemalloc
 endif