--- cat ebuild.diff | sed s/EBUILD/mplayer-1.0.20070814.ebuild/ | patch -p0
+++ EBUILD
@@ -5,7 +5,7 @@
 inherit eutils flag-o-matic multilib
 
 RESTRICT="strip"
-IUSE="3dnow 3dnowext a52 aac aalib alsa altivec amrnb amrwb arts bidi bl bindist
+IUSE="darknrg 3dnow 3dnowext a52 aac aalib alsa altivec amrnb amrwb arts bidi bl bindist
 cddb cpudetection custom-cflags dga doc dts dvb cdparanoia directfb dvd
 dvdnav dv enca encode esd fbcon ftp gif ggi gtk iconv ipv6 ivtv jack joystick
 jpeg libcaca lirc live livecd lzo mad md5sum mmx mmxext mp2 mp3 musepack nas
@@ -189,7 +189,9 @@
 
 	# Fix XShape detection
 	epatch ${FILESDIR}/${PN}-xshape.patch
	epatch ${FILESDIR}/${PN}-dpms.patch
+
+	use darknrg && epatch ${FILESDIR}/darknrg.patch
 }
 
 src_compile() {
