--- any2vob	2007-05-26 14:48:26.000000000 +0200
+++ any2vob_patched	2007-05-26 14:48:17.000000000 +0200
@@ -1014,7 +1014,7 @@
 	fi
 	[ "$SKIP_VOB" = "yes" ] && return
 	surround_check
-	if [ "$TRANSCODE_VERSION" \> "1.0.2" ]; then
+	if [ "$TRANSCODE_VERSION" \= "1.0.2" ]; then
 		PROGRESS_ARGS="--progress_meter 0"
 	else
 		PROGRESS_ARGS="--progress_off"
@@ -1921,7 +1921,7 @@
 	[ "$TOTAL" -lt 1 ] && TOTAL=1
 	PROCESS_FILE="$WORKDIR"/"$name".m2v
 	preview
-	if [ "$TRANSCODE_VERSION" \> "1.0.2" ]; then
+	if [ "$TRANSCODE_VERSION" \= "1.0.2" ]; then
 		outfile="$name".m2v
 		tmp_outfile="$name"_tmp.m2v
 		PROGRESS_ARGS="--progress_meter 0"
