--- any2dvd-0.34/any2dvd.orig	2007-04-24 17:10:06.000000000 -0500
+++ any2dvd-0.34/any2dvd	2007-04-24 17:10:50.000000000 -0500
@@ -733,7 +733,7 @@
 
 menu_audio() {						# Create audio for VMGM/VTSM Menus
 	if [ ! -z "$VMGMAUD" ]; then
-		ffmpeg -y -i "$VMGMAUD" -ar 48000 -ab 384 -ac 2 "$WORKDIR"/audio/vmgm-audio.ac3 &> "$WORKDIR"/tmp.log &
+		ffmpeg -y -i "$VMGMAUD" -ar 48000 -ab 384k -ac 2 "$WORKDIR"/audio/vmgm-audio.ac3 &> "$WORKDIR"/tmp.log &
 		echo $! > "$WORKDIR"/pids.dat
 		sleep 5s
 		IFS=$REAL_IFS
@@ -749,7 +749,7 @@
 		"
 	else
 		ecasound -i null -t:20 -o "$WORKDIR"/audio/vmgm-audio.wav &> /dev/null
-		ffmpeg -y -i "$WORKDIR"/audio/vmgm-audio.wav -ar 48000 -ab 224 -ac 2 "$WORKDIR"/audio/vmgm-audio.ac3 &> "$WORKDIR"/tmp.log &
+		ffmpeg -y -i "$WORKDIR"/audio/vmgm-audio.wav -ar 48000 -ab 224k -ac 2 "$WORKDIR"/audio/vmgm-audio.ac3 &> "$WORKDIR"/tmp.log &
 		echo $! > "$WORKDIR"/pids.dat
 		sleep 5s
 		IFS=$REAL_IFS
@@ -766,7 +766,7 @@
 		rm "$WORKDIR"/audio/vmgm-audio.wav
 	fi
 	if [ ! -z "$VTSMAUD" ]; then
-		ffmpeg -y -i "$VTSMAUD" -ar 48000 -ab 384 -ac 2 "$WORKDIR"/audio/vtsm-audio.ac3 &> "$WORKDIR"/tmp.log &
+		ffmpeg -y -i "$VTSMAUD" -ar 48000 -ab 384k -ac 2 "$WORKDIR"/audio/vtsm-audio.ac3 &> "$WORKDIR"/tmp.log &
 		echo $! > "$WORKDIR"/pids.dat
 		sleep 5s
 		IFS=$REAL_IFS
@@ -782,7 +782,7 @@
 		"
 	else
 		ecasound -i null -t:20 -o "$WORKDIR"/audio/vtsm-audio.wav &> /dev/null
-		ffmpeg -y -i "$WORKDIR"/audio/vtsm-audio.wav -ar 48000 -ab 224 -ac 2 "$WORKDIR"/audio/vtsm-audio.ac3 &> "$WORKDIR"/tmp.log &
+		ffmpeg -y -i "$WORKDIR"/audio/vtsm-audio.wav -ar 48000 -ab 224k -ac 2 "$WORKDIR"/audio/vtsm-audio.ac3 &> "$WORKDIR"/tmp.log &
 		echo $! > "$WORKDIR"/pids.dat
 		sleep 5s
 		IFS=$REAL_IFS
