--- any2vob-0.36_beta-r2.sh	2007-10-07 15:10:36.000000000 -0500
+++ any2vob-0.36_beta2.sh	2007-10-17 19:12:58.000000000 -0500
@@ -514,7 +514,7 @@
 	TRANSCODE_PID="$!"
 	# Transcode sometimes has problems exiting cleanly with '-c', hard kill it manually if it hasn't exited after 15
 	sleep 15s
-	if [ ! -z "`ps aux | grep " $TRANSCODE_PID "`" ]; then
+	if [ ! -z "`ps aux | grep " $TRANSCODE_PID " | grep -v grep`" ]; then
 		kill -9 "$TRANSCODE_PID" &> /dev/null
 	fi
 	PROGRESSIVE=`grep progressive "$WORKDIR"/fieldanalyse.dat | wc -l`
@@ -2087,7 +2087,7 @@
 	if [ ! -z "$RESAMPLE_ARGS" ]; then
 		echo -en "\n\e[33;01m Checking if Transcode can decode audio stream ...\e[m"
 		eval transcode $PROBE_ARGS -i \"$file\" $IMPORT_ARGS $EXPORT_ARGS $ASPECT_ARGS $TRANSCODE_CROP_ARGS $FILTER_ARGS $AUDIO_ARGS -c 0-200 -o \""$WORKDIR"/"$test_outfile"\" -w $VID_BITRATE &> /dev/null
-		sleep 15s && [ ! -z "`ps aux | grep " $TRANSCODE_PID "`" ] && kill -9 "$TRANSCODE_PID" &> /dev/null	# Transcode sometimes has problems exiting cleanly with '-c', hard kill it manually if it hasn't exited after 15 seconds
+		sleep 15s && [ ! -z "`ps aux | grep " $TRANSCODE_PID " | grep -v grep`" ] && kill -9 "$TRANSCODE_PID" &> /dev/null	# Transcode sometimes has problems exiting cleanly with '-c', hard kill it manually if it hasn't exited after 15 seconds
 		if [ ! -s "$WORKDIR"/transcode_test.m2v ]; then
 			RESAMPLE_ARGS=
 			TRANSCODE_BAD_AUDIO="yes"
@@ -2108,7 +2108,7 @@
 		echo -en "\e[33;01m Checking if Transcode can decode video stream without MPlayer ...\e[m"
 		eval transcode -i \"$file\" $EXPORT_ARGS $ASPECT_ARGS $TRANSCODE_CROP_ARGS $FILTER_ARGS $AUDIO_ARGS -c 0-0:0:5 &> "$WORKDIR"/transcode_test &
 		TRANSCODE_PID="$!"
-		sleep 15s && [ ! -z "`ps aux | grep " $TRANSCODE_PID "`" ] && kill -9 "$TRANSCODE_PID" &> /dev/null	# Transcode sometimes has problems exiting cleanly with '-c', hard kill it manually if it hasn't exited after 15 seconds
+		sleep 15s && [ ! -z "`ps aux | grep " $TRANSCODE_PID " | grep -v grep`" ] && kill -9 "$TRANSCODE_PID" &> /dev/null	# Transcode sometimes has problems exiting cleanly with '-c', hard kill it manually if it hasn't exited after 15 seconds
 		if [ ! -z "`grep -E \"V=(null)|A=(null)|/dev/zero|(module=null)\" "$WORKDIR"/transcode_test`" ]; then
 			TRANSCODE_BAD_DECODE="yes"	# Check to see if Transcode can decode by itself
 			echo -e "\e[33;01m NO\e[m\n"
@@ -2156,13 +2156,13 @@
 			# Tested against 1.1.0 CVS
 			tail -fn1 "$WORKDIR"/tmp.log &
 			TAIL_PID="$!"
-			while [ ! -z "`ps aux | grep " $TRANSCODE_PID "`" ]; do
+			while [ ! -z "`ps aux | grep " $TRANSCODE_PID " | grep -v grep`" ]; do
 				sleep 1s
 			done
 			kill "$TAIL_PID" &> /dev/null
 			echo
 		else
-			while [ ! -z "`ps aux | grep " $TRANSCODE_PID "`" ]; do
+			while [ ! -z "`ps aux | grep " $TRANSCODE_PID " | grep -v grep`" ]; do
 				TRANSCODE_PROC=`tail -n1 "$WORKDIR"/tmp.log`
 				if [ ! -z "$TRANSCODE_PROC" ]; then
 					if [ "$TERM" = "xterm" ]; then	# Non xterm's scroll up one line when issuing backspaces after reaching the leftmost limit
