--- configure.ac.org	2009-03-26 18:15:43.000000000 +0100
+++ configure.ac	2009-03-26 18:15:59.000000000 +0100
@@ -379,12 +379,12 @@
 	if test "$withval" = "no"; then
		BASH_COMPLETION_DIR=""
 	elif test "$withval" = "yes"; then
-		BASH_COMPLETION_DIR=/etc/bash_completion.d
+		BASH_COMPLETION_DIR='${DESTDIR}'/etc/bash_completion.d
 	else
-		BASH_COMPLETION_DIR="$withval"
+		BASH_COMPLETION_DIR='${DESTDIR}'"${withval}"
 	fi
 	],
-	[BASH_COMPLETION_DIR=/etc/bash_completion.d]
+	[BASH_COMPLETION_DIR='${DESTDIR}'/etc/bash_completion.d]
 )
 AC_SUBST(BASH_COMPLETION_DIR)
 
--- configure.org	2009-03-26 18:25:33.000000000 +0100
+++ configure	2009-03-26 18:26:16.000000000 +0100
@@ -6933,15 +6933,15 @@
 if test "${with_bash_completion+set}" = set; then
   withval=$with_bash_completion;
 	if test "$withval" = "no"; then
-		BASH_COMPLETION_DIR=""
+		BASH_COMPLETION_DIR=""
 	elif test "$withval" = "yes"; then
-		BASH_COMPLETION_DIR=/etc/bash_completion.d
+		BASH_COMPLETION_DIR='${DESTDIR}'/etc/bash_completion.d
 	else
-		BASH_COMPLETION_DIR="$withval"
+		BASH_COMPLETION_DIR='${DESTDIR}'"$withval"
 	fi
 
 else
-  BASH_COMPLETION_DIR=/etc/bash_completion.d
+  BASH_COMPLETION_DIR='${DESTDIR}'/etc/bash_completion.d
 
 fi
 
--- Makefile.in.org	2009-03-26 18:40:19.000000000 +0100
+++ Makefile.in	2009-03-26 18:41:11.000000000 +0100
@@ -31,6 +31,7 @@
 	@rm -f ${DESTDIR}${prefix}/bin/ptxdist
 	@ln -sf ${instdir}/bin/ptxdist ${DESTDIR}${prefix}/bin/ptxdist-${version}
 	@ln -sf ${instdir}/bin/ptxdist ${DESTDIR}${prefix}/bin/ptxdist
+	@mkdir -p "@BASH_COMPLETION_DIR@"
 	-@[ -d "@BASH_COMPLETION_DIR@" ] && cp scripts/bash_completion "@BASH_COMPLETION_DIR@/ptxdist"
 
 clean:
