diff -Naur src.old/KnockOutGame.cpp src/KnockOutGame.cpp
--- src.old/KnockOutGame.cpp	2008-09-30 10:22:40.000000000 +0200
+++ src/KnockOutGame.cpp	2008-09-30 10:19:52.000000000 +0200
@@ -7,6 +7,7 @@
  *
  */
 
+#include <cstdio>
 #include "KnockOutGame.h"
 #include "GameConstants.h"
 #include "ScException.h"
diff -Naur src.old/Player.cpp src/Player.cpp
--- src.old/Player.cpp	2008-09-30 10:22:40.000000000 +0200
+++ src/Player.cpp	2008-09-30 10:18:01.000000000 +0200
@@ -14,6 +14,7 @@
  *  Created by Ruben Henner Zilibowitz on 2/01/05.
  */
 
+#include <cstdio>
 #include "Player.h"
 #include "CarGame.h"
 
diff -Naur src.old/RobotHotPotatoPlayer.cpp src/RobotHotPotatoPlayer.cpp
--- src.old/RobotHotPotatoPlayer.cpp	2008-09-30 10:22:40.000000000 +0200
+++ src/RobotHotPotatoPlayer.cpp	2008-09-30 10:20:34.000000000 +0200
@@ -14,6 +14,7 @@
  *  Created by Ruben Henner Zilibowitz on 31/10/07.
  */
 
+#include <cstdio>
 #include "RobotHotPotatoPlayer.h"
 #include "HotPotatoGame.h"
 #include <cassert>
diff -Naur src.old/Startline.cpp src/Startline.cpp
--- src.old/Startline.cpp	2008-09-30 10:22:40.000000000 +0200
+++ src/Startline.cpp	2008-09-30 10:18:59.000000000 +0200
@@ -14,6 +14,7 @@
  *  Created by Ruben Henner Zilibowitz on 19/04/05.
  */
 
+#include <cstdio>
 #include "Startline.h"
 #include "ScException.h"
 #include <algorithm>
