--- /usr/local/portage/dev-util/atprepare/files/atprepare.sh	2007-09-19 00:42:42.000000000 +0200
+++ atprepare	2007-11-06 18:23:30.000000000 +0100
@@ -8,7 +8,7 @@
 function create_makefile_src() { # skip
 	echo "Creating source Makefile.am ..."
 	SOURCES=""
-	for f in *.[ch] *.cpp *.cxx *.hh; do
+	for f in *.c *.cpp *.cxx; do
 		if [ -f "$f" ]; then
 			[ "$f" != "${f%.c}" ] && HAS_C=1
 			[ "$f" != "${f%.cpp}" ] && HAS_CXX=1
@@ -103,8 +103,9 @@
 config.status
 configure
 stamp-h1
-$PROGNAME
-" >> .cvsignore
+$PROGNAME" >> .cvsignore
+sort .cvsignore | uniq > .cvsignore.tmp
+mv -f .cvsignore.tmp .cvsignore
 }
 
 function create_bootstrap() {
@@ -122,10 +123,15 @@
 chmod 755 bootstrap
 }
 
+[ "`whoami`" == "root" ] && {
+	echo "Error: Should not be run as root!"
+	exit 1
+}
+
 [ "$1" == "remove" ] && {
 	rm -f -r autom4te.cache .deps
 	rm -f aclocal.m4 autoscan-*.log bootstrap config.h config.h.in config.log config.status configure
-	rm -f configure.in configure.scan depcomp install-sh missing stamp-h1
+	rm -f configure.in configure.scan depcomp install-sh missing stamp-h1 config.h.in~
 	rm -f Makefile.am Makefile.in Makefile
 	[ -d "src" ] && {
 		rm -f -r src/.deps 
@@ -134,18 +140,44 @@
 	exit 
 }
 
+[ "$1" == "dist" ] && {
+	[ "$2" ] && ver=$2 || ver="1.0"
+	echo "Checking out in temp dir ..."
+	tmp=`mktemp -d`
+	olddir=`pwd`
+	subdir=`basename $olddir`
+	svnrepos=`grep svn .svn/entries | head -n 1`
+	trg="/tmp/$subdir-$ver.tar.bz2"
+	cd $tmp
+	svn checkout $svnrepos
+	cd $subdir
+	echo "Removing cvs/svn information ..."
+	find . -type d -name ".svn" -print0 -o -type d -name "CVS" -print0 | xargs -0r rm -f -r
+	rm -f .cvsignore
+	echo "Tarring archive ..."
+	cd ..
+	tar cv $subdir | bzip2 -c -f -9 > $trg
+	echo "Cleaning up ..."
+	cd $olddir
+	rm -fr $tmp
+	echo "Done - archive is in $trg"
+	exit
+}
+
 [ "$1" != "prep" ] && {
-	echo "Syntax: $0 prep|remove"
+	echo "Syntax: $0 prep|remove|dist"
 	echo "prep:   prepare a C or C++ project with existing source files in . or in src/"
 	echo "        for autotools usage."
 	echo "remove: remove all autotools relevant files from the current directory and"
 	echo "        the src/ subdirectory."
+	echo "dist:   create tar.bz2 archive for distributioning"
 	exit 1
 }
 
 
 WD=`pwd`
 PROGNAME=`basename $WD`
+PROGNAME=${PROGNAME//-/_}
 rm -f makefile Makefile configure.in
 [ -e "Makefile.am" ] && skip="skip" || skip=""
 if [ -d "src" ]; then
