--- glibc-2.7/sysdeps/i386/i486/bits/string.h.org	2007-09-15 02:36:23.000000000 +0000
+++ glibc-2.7/sysdeps/i386/i486/bits/string.h		2007-10-23 09:52:11.126980091 +0000
@@ -143,10 +143,13 @@

 #define _HAVE_STRING_ARCH_memmove 1
 #ifndef _FORCE_INLINES
+#define memmove(dest, src, n) \
+__memmove_g ((dest), (src), (n))
+
 /* Copy N bytes of SRC to DEST, guaranteeing
    correct behavior for overlapping strings.  */
 __STRING_INLINE void *
-memmove (void *__dest, __const void *__src, size_t __n)
+__memmove_g (void *__dest, __const void *__src, size_t __n)
 {
   register unsigned long int __d0, __d1, __d2;
   register void *__tmp = __dest; 
