diff -r -U3 kipi-plugins-0.1.3/kipi-plugins/rawconverter/dcrawbinary.cpp kipi-plugins-0.1.3.new/kipi-plugins/rawconverter/dcrawbinary.cpp
--- kipi-plugins-0.1.3/kipi-plugins/rawconverter/dcrawbinary.cpp	2007-01-24 22:04:18.000000000 +0100
+++ kipi-plugins-0.1.3.new/kipi-plugins/rawconverter/dcrawbinary.cpp	2007-02-04 11:58:33.000000000 +0100
@@ -93,7 +93,7 @@
 
 const char *DcrawBinary::path()
 {
-    return "kipidcraw";
+    return "dcraw";
 }
 
 bool DcrawBinary::isAvailable() const
diff -r -U3 kipi-plugins-0.1.3/kipi-plugins/rawconverter/Makefile.am kipi-plugins-0.1.3.new/kipi-plugins/rawconverter/Makefile.am
--- kipi-plugins-0.1.3/kipi-plugins/rawconverter/Makefile.am	2007-01-24 22:04:18.000000000 +0100
+++ kipi-plugins-0.1.3.new/kipi-plugins/rawconverter/Makefile.am	2007-02-04 12:01:30.000000000 +0100
@@ -35,24 +35,3 @@
 messages: rc.cpp
 	$(XGETTEXT) *.cpp *.h -o $(podir)/kipiplugin_rawconverter.pot
 
-#-- dcraw binary -------------------------------------------------
-
-# RawConverter plugin use a dedicaced dcraw binary program to run. This is mandatory 
-# since dcraw author have broken the command line options 
-# compatibility with the 8.x serie.
-# Also, dcraw is not available like a library.
-# The dcraw.c source code is just a copy of official implementation, but 
-# this one have been completly tested with this plugin. 
-# Note: dcraw project page is http://cybercom.net/~dcoffin
-
-bin_PROGRAMS = kipidcraw
-
-# This line is require to disable color management support (using LCMS library) witch 
-# is not yet supported by plugin.
-kipidcraw_CFLAGS = -DNO_LCMS -w 
-
-kipidcraw_SOURCES = dcraw.c
-
-kipidcraw_LDFLAGS = $(all_libraries) $(LIBJPEG) -lm
-
-
