--- install.sh.old	2007-08-31 01:52:25.000000000 +0200
+++ install.sh	2007-10-02 13:19:00.000000000 +0200
@@ -33,16 +33,22 @@
 fi
 
 VERSION="2.0-RC6"
-BONGO_INST="/usr/share/bongosurfer"
-BONGO_SKRIPTE="${BONGO_INST}/skripte"
-BONGO_JAR="${BONGO_INST}/jar"
-BONGO_LIB="${BONGO_INST}/lib"
-BONGO_DOC="/usr/share/doc/bongosurfer"
-BONGO_CONF="/etc/bongosurfer"
-MODEM="${BONGO_SKRIPTE}/modem"
-I4L="${BONGO_SKRIPTE}/i4l"
-CAPI="${BONGO_SKRIPTE}/capi"
-TIME="${BONGO_SKRIPTE}/time"
+[ -z "${BASEDIR}" ]        	&& BASEDIR=/
+[ -z "${PREFIX}" ]        	&& PREFIX=/usr
+[ -z "${GLOBAL_CONF}" ]   && GLOBAL_CONF=/etc
+[ -z "${BONGO_CONF}" ] && BONGO_CONF="${BASEDIR}${GLOBAL_CONF}/bongosurfer"
+[ -z "${BONGO_INST}" ]       && BONGO_INST="${BASEDIR}${PREFIX}/share/bongosurfer"
+[ -z "${BONGO_DOC}" ] && BONGO_DOC="${BASEDIR}${PREFIX}/share/doc/bongosurfer"
+[ -z "${BONGO_SKRIPTE}" ] && BONGO_SKRIPTE="${BONGO_INST}/skripte"
+[ -z "${BONGO_JAR}" ] && BONGO_JAR="${BONGO_INST}/jar"
+[ -z "${BONGO_LIB}" ] && BONGO_LIB="${BONGO_INST}/lib"
+[ -z "${BONGO_MAN}" ] && BONGO_MAN="${BASEDIR}${PREFIX}/share/man/man1"
+[ -z "${BONGO_ICON}" ] && BONGO_ICON="${BASEDIR}${PREFIX}/share/pixmaps"
+[ -z "${BONGO_APPLNK}" ] && BONGO_APPLNK="${BASEDIR}${PREFIX}/share/applnk"
+[ -z "${MODEM}" ] && MODEM="${BONGO_SKRIPTE}/modem"
+[ -z "${I4L}" ] && I4L="${BONGO_SKRIPTE}/i4l"
+[ -z "${CAPI}" ] && CAPI="${BONGO_SKRIPTE}/capi"
+[ -z "${TIME}" ] && TIME="${BONGO_SKRIPTE}/time"
 
 PATH=$PATH:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/usr/local/sbin:/bin
 
@@ -112,6 +118,18 @@
 	|| echo "Java ist nicht installiert oder nicht im Suchpfad -> Wenn Java schon installiert
 ist, java-/bin/-Verzeichnis in PATH Variable aufnehmen."
 
+if [ ! -d ${BASEDIR}${PREFIX}/bin ]; then
+	install -vm 755 -d ${BASEDIR}${PREFIX}/bin
+fi
+if [ ! -d ${BASEDIR}${PREFIX}/sbin ]; then
+	install -vm 755 -d ${BASEDIR}${PREFIX}/sbin
+fi
+if [ ! -d ${BONGO_MAN} ]; then
+	install -vm 755 -d ${BONGO_MAN}
+fi
+if [ ! -d ${BONGO_ICON} ]; then
+	install -vm 755 -d ${BONGO_ICON}
+fi
 if [ ! -d ${BONGO_INST} ]; then
 	install -vm 755 -d ${BONGO_INST}
 fi
@@ -127,46 +145,46 @@
 install -vm 755 -d ${CAPI}
 install -vm 755 -d ${TIME}
 
-install -vm 700 -d /etc/ppp/peers/bongosurfer
+install -vm 700 -d ${BASEDIR}${GLOBAL_CONF}/ppp/peers/bongosurfer
 
 install -vm 644 ./doc/COPYING ${BONGO_INST}
 install -vm 644 ./doc/* ${BONGO_DOC}
 
-install --group=dialout -vm 750 ./skripte/bin/bongosurfer /usr/bin/bongo
+install --group=dialout -vm 750 ./skripte/bin/bongosurfer ${BASEDIR}${PREFIX}/bin/bongo
 
 install -vm 644 ./icons/bongosurfer.png ${BONGO_INST}
 install -vm 755 ./skripte/modem/* ${MODEM}
 install -vm 755 ./skripte/i4l/* ${I4L}
 install -vm 755 ./skripte/capi/* ${CAPI}
 
-install -vm 744 ./peers/* /etc/ppp/peers/bongosurfer
-install -vm 700 ./skripte/sbin/bongosetup /usr/sbin/bongosetup2
+install -vm 744 ./peers/* ${BASEDIR}${GLOBAL_CONF}/ppp/peers/bongosurfer
+install -vm 700 ./skripte/sbin/bongosetup ${BASEDIR}${PREFIX}/sbin/bongosetup2
 install --group=dialout -vm 750 ./skripte/time/synchronizeClock.sh ${TIME}
 
 install -vm 644 ./jar/*.jar ${BONGO_JAR}
 install -vm 644 ./lib/*.jar ${BONGO_LIB}
 
-install -vm 644 ./man/*.gz /usr/share/man/man1/
+install -vm 644 ./man/*.gz ${BONGO_MAN}/
 
 if [ -d /usr/share/applnk/Internet ]; then
-    install -vm 644 ./apps/bongosurfer.desktop /usr/share/applnk/Internet/
+    install -vm 644 ./apps/bongosurfer.desktop ${BONGO_APPLNK}/Internet/
 elif [ -d /usr/share/applnk/Utilities ]; then
-    install -vm 644 ./apps/bongosurfer.desktop /usr/share/applnk/Utilities/
+    install -vm 644 ./apps/bongosurfer.desktop ${BONGO_APPLNK}/Utilities/
 elif [ -d /usr/share/applnk/Applications ]; then
-    install -vm 644 ./apps/bongosurfer.desktop /usr/share/applnk/Applications/
+    install -vm 644 ./apps/bongosurfer.desktop ${BONGO_APPLNK}/Applications/
 fi
 
-install -vm 644 icons/bongosurfer.png /usr/share/pixmaps/
+install -vm 644 icons/bongosurfer.png ${BONGO_ICON}/
 
 for datei in config/*; do
 	ConfigDatei=`basename $datei`
 	installConfigDatei
 done
 
-chmod a+rx /etc/ppp/peers/bongosurfer
-chmod a+r /etc/ppp/peers/bongosurfer/*
+chmod a+rx ${BASEDIR}${GLOBAL_CONF}/ppp/peers/bongosurfer
+chmod a+r ${BASEDIR}${GLOBAL_CONF}/ppp/peers/bongosurfer/*
 
-/usr/sbin/bongosetup2 -a
+#/usr/sbin/bongosetup2 -a
 echo
 echo
 echo "Installation von BongoSurfer abgeschlossen."
