--- src.org/DownloadListCtrl.cpp
+++ src/DownloadListCtrl.cpp
@@ -576,7 +576,16 @@ void CDownloadListCtrl::OnCancelFile(wxC
 	ItemList files = ::GetSelectedItems(this, itFILES);
 	if (files.size()) {	
 		wxString question = 
-			_("Are you sure that you wish to delete the selected file(s)?");	
+			_("Are you sure that you wish to delete the selected file(s):\n");
+		int cnt = 0;
+		for (ItemList::iterator it = files.begin(); it != files.end(); ++it) {
+			if (++cnt == 5) {
+				question += wxT("...");
+				break;
+			}
+			CPartFile* file = (*it)->GetFile();
+			question += file->GetFileName() + wxT("\n");
+		}
 		if (wxMessageBox( question, _("Cancel"), wxICON_QUESTION | wxYES_NO, this) == wxYES) {
 			for (ItemList::iterator it = files.begin(); it != files.end(); ++it) {
 				CPartFile* file = (*it)->GetFile();
