--- src.org/DownloadListCtrl.cpp
+++ src/DownloadListCtrl.cpp
@@ -576,9 +576,18 @@ void CDownloadListCtrl::OnCancelFile(wxC
 	if (files.size()) {	
 		wxString question;
 		if (files.size() == 1) {
-			question = _("Are you sure that you wish to delete the selected file?");
+			question = _("Are you sure that you wish to delete the selected file:\n");
 		} else {
-			question = _("Are you sure that you wish to delete the selected files?");
+			question = _("Are you sure that you wish to delete the selected files:\n");
+		}
+		int cnt = 0;
+		for (ItemList::iterator it = files.begin(); it != files.end(); ++it) {
+			if (++cnt == 5) {
+				question += wxT("...");
+				break;
+			}
+			CPartFile* file = (*it)->GetFile();
+			question += file->GetFileName().GetPrintable() + wxT("\n");
 		}
 		if (wxMessageBox( question, _("Cancel"), wxICON_QUESTION | wxYES_NO, this) == wxYES) {
 			for (ItemList::iterator it = files.begin(); it != files.end(); ++it) {
