--- src.org/DownloadListCtrl.cpp
+++ src/DownloadListCtrl.cpp
@@ -154,6 +154,7 @@ BEGIN_EVENT_TABLE(CDownloadListCtrl, CMu
 	EVT_MENU( MP_METINFO,			CDownloadListCtrl::OnViewFileInfo )
 	EVT_MENU( MP_VIEW,			CDownloadListCtrl::OnPreviewFile )
 	EVT_MENU( MP_VIEWFILECOMMENTS,		CDownloadListCtrl::OnViewFileComments )
+	EVT_MENU( MP_VIEWALLSHAREDFILES,		CDownloadListCtrl::OnViewAllSharedFiles)
 
 	EVT_MENU( MP_WS,			CDownloadListCtrl::OnGetFeedback )
 	EVT_MENU( MP_RAZORSTATS, 		CDownloadListCtrl::OnGetRazorStats )
@@ -840,14 +841,33 @@ void CDownloadListCtrl::OnViewFiles( wxC
 {
 	ItemList sources = ::GetSelectedItems( this, itSOURCES );
 
-	if ( sources.size() == 1 ) {
-		CUpDownClient* source = sources.front()->GetSource();
-		
+	for ( ItemList::iterator it = sources.begin(); it != sources.end(); ++it ) {
+		CUpDownClient* source = (*it)->GetSource();
+
 		source->RequestSharedFileList();
 	}
 }
 
 
+void CDownloadListCtrl::OnViewAllSharedFiles( wxCommandEvent& WXUNUSED(event) )
+{
+	ItemList files = ::GetSelectedItems( this, itFILES );
+
+	for ( ItemList::iterator it = files.begin(); it != files.end(); ++it ) {
+		CPartFile* file = (*it)->GetFile();
+
+		const CPartFile::SourceSet& normSources = file->GetSourceList();
+
+		CPartFile::SourceSet::iterator it2;
+		for ( it2 = normSources.begin(); it2 != normSources.end(); ++it2 ) {
+			CUpDownClient* source = (*it2);
+
+			source->RequestSharedFileList();
+		}
+	}
+}
+
+
 void CDownloadListCtrl::OnAddFriend( wxCommandEvent& WXUNUSED(event) )
 {
 	ItemList sources = ::GetSelectedItems( this, itSOURCES );
@@ -981,6 +1001,7 @@ void CDownloadListCtrl::OnMouseRightClic
 		m_menu->Append(MP_METINFO, _("Show file &details"));
 		m_menu->Append(MP_VIEWFILECOMMENTS,
 			_("Show all comments"));
+		m_menu->Append(MP_VIEWALLSHAREDFILES,_("View all shared files"));
 		//-----------------------------------------------------
 		m_menu->AppendSeparator();
 		//-----------------------------------------------------
--- src.org/DownloadListCtrl.h
+++ src/DownloadListCtrl.h
@@ -230,6 +230,7 @@ private:
 	void	OnViewFileInfo( wxCommandEvent& event );
 	void	OnViewFileComments( wxCommandEvent& event );
 	void	OnPreviewFile( wxCommandEvent& event );
+	void	OnViewAllSharedFiles( wxCommandEvent& event );
 	
 	// Event-handlers for sources
 	void	OnSwapSource( wxCommandEvent& event );
--- src.org/include/common/MenuIDs.h
+++ src/include/common/MenuIDs.h
@@ -71,6 +71,7 @@ enum {
 	MP_ADDTOSTATIC,
 	MP_REMOVEFROMSTATIC,
 	MP_VIEWFILECOMMENTS,
+	MP_VIEWALLSHAREDFILES,
 	MP_CAT_ADD,
 	MP_CAT_EDIT,
 	MP_CAT_REMOVE,
