--- src.org/MuleNotebook.cpp
+++ src/MuleNotebook.cpp
@@ -66,6 +66,8 @@ bool CMuleNotebook::DeletePage(int nPage
 	wxCHECK_MSG((nPage >= 0) && (nPage < (int)GetPageCount()), false,
 		wxT("Trying to delete invalid page-index in CMuleNotebook::DeletePage"));
 
+	return false; // amule crash on page close with older wxGTK versions
+
 	// Send out close event
 	wxNotebookEvent evt( wxEVT_COMMAND_MULENOTEBOOK_PAGE_CLOSING, GetId(), nPage );
 	evt.SetEventObject(this);
@@ -95,7 +97,8 @@ bool CMuleNotebook::DeleteAllPages()
 	Freeze();
 
 	bool result = true;
-	while ( GetPageCount() ) {
+	// avoid deadlock if DeletePage() returns false
+	while ( result && GetPageCount() ) {
 		result &= DeletePage( 0 );
 	}
 	
