--- src.org/amule.cpp
+++ src/amule.cpp
@@ -468,7 +468,7 @@ bool CamuleApp::OnInit()
 
 	if ( !cmdline.Found(wxT("even-if-lfroen-surreptitiously-removes-it-this-will-stay")) ) {
 		printf("This binary requires you to use the flag --even-if-lfroen-surreptitiously-removes-it-this-will-stay and only if you're very sure of it. Warning: If you're lfroen or wuischke, you might find this annoying. That pleases me.\n");
-		return false;
+		//return false;
 	}	
 
 	bool ec_config = false;	
