diff -BuNr orig/pcsx2/Interpreter.c src/pcsx2/Interpreter.c
--- orig/pcsx2/Interpreter.c	2007-04-01 22:26:47.000000000 -0400
+++ src/pcsx2/Interpreter.c	2007-08-14 05:34:48.000000000 -0400
@@ -959,7 +959,9 @@
 
 extern void iDumpVU0Registers();
 extern void iDumpVU1Registers();
+#ifdef PCSX2_DEVBUILD
 extern u32 vudump;
+#endif
 extern int vu0branch, vu1branch;
 
 void intExecuteVU0Block() {
diff -BuNr orig/pcsx2/Linux/Config.c src/pcsx2/Linux/Config.c
--- orig/pcsx2/Linux/Config.c	2007-04-01 22:26:37.000000000 -0400
+++ src/pcsx2/Linux/Config.c	2007-08-14 05:01:15.000000000 -0400
@@ -91,7 +91,10 @@
 	GetValuel("SafeCnts", Config.SafeCnts);
 
 	GetValuel("Patch",      Config.Patch);
+
+#ifdef PCSX2_DEVBUILD
 	GetValuel("varLog", varLog);
+#endif
 	
 	free(data);
 
@@ -138,7 +141,9 @@
 	SetValuel("SafeCnts", Config.SafeCnts);
 	// misc
 	SetValuel("Patch",      Config.Patch);
+#ifdef PCSX2_DEVBUILD
 	SetValuel("varLog", varLog);
+#endif
 
 	fclose(f);
 
diff -BuNr orig/pcsx2/Linux/GtkGui.c src/pcsx2/Linux/GtkGui.c
--- orig/pcsx2/Linux/GtkGui.c	2007-04-01 22:26:37.000000000 -0400
+++ src/pcsx2/Linux/GtkGui.c	2007-08-14 05:15:05.000000000 -0400
@@ -1208,8 +1208,11 @@
 	UpdateDebugger();
 }
 
+
 void OnDebug_Log(GtkButton *button, gpointer user_data) {
-	Log = 1 - Log;
+#ifdef PCSX2_DEVBUILD
+		Log = 1 - Log;
+#endif
 }
 
 void OnDebug_EEMode(GtkToggleButton *togglebutton, gpointer user_data) {
@@ -1293,6 +1296,8 @@
 }
 
 void OnLogging_Ok(GtkButton *button, gpointer user_data) {
+
+#ifdef PCSX2_DEVBUILD
 	GtkWidget *Btn;
 	char str[32];
 	int i, ret;
@@ -1323,6 +1328,7 @@
 
 	Btn = lookup_widget(LogDlg, "Log");
 	Log = gtk_toggle_button_get_active(GTK_TOGGLE_BUTTON(Btn));
+#endif
 
 	SaveConfig();
 
diff -BuNr orig/pcsx2/Linux/LnxMain.c src/pcsx2/Linux/LnxMain.c
--- orig/pcsx2/Linux/LnxMain.c	2007-04-01 22:26:37.000000000 -0400
+++ src/pcsx2/Linux/LnxMain.c	2007-08-14 05:26:46.000000000 -0400
@@ -35,7 +35,6 @@
 #include <glib/gthread.h>
 #include <unistd.h>
 
-
 #include "Common.h"
 #include "PsxCommon.h"
 #include "Linux.h"
@@ -112,13 +111,13 @@
 				"\t-numimages [num] {after hitting frame, this many images will be captures every 20 frames}\n"
                 "\t-test {Triggers testing mode (only for dev builds)}\n"
                 "\n"
-#endif
 
                 "Load Plugins:\n"
                 "\t-cdvd [dllpath] {specify the dll load path of the CDVD plugin}\n"
                 "\t-gs [dllpath] {specify the dll load path of the GS plugin}\n"
                 "-pad [tsxcal] {specify to hold down on the triangle, square, circle, x, start, select buttons}\n"
                 "\t-spu [dllpath] {specify the dll load path of the SPU2 plugin}\n"
+#endif
                 "\n";
 
 			//SysMessage(phelpmsg);
@@ -132,62 +131,65 @@
 		}
         else if( stricmp(token, "-nogui") == 0 ) {
             UseGui = 0;
+
         }
-		else if( stricmp(token, "-image") == 0 ) {
-			g_TestRun.pimagename = argv[i++];
-		}
-		else if( stricmp(token, "-loadgs") == 0 ) {
-			g_pRunGSState = argv[i++];
-		}
-		else if( stricmp(token, "-log") == 0 ) {
-			g_TestRun.plogname = argv[i++];
-		}
-		else if( stricmp(token, "-logopt") == 0 ) {
-			token = argv[i++];
-			if( token != NULL ) {
-				if( token[0] == '0' && token[1] == 'x' ) token += 2;
-				sscanf(token, "%x", &varLog);
-			}
-		}
-		else if( stricmp(token, "-pad") == 0 ) {
-			token = argv[i++];
-			printf("-pad ignored\n");
-		}
-		else if( stricmp(token, "-frame") == 0 ) {
-			token = argv[i++];
-			if( token != NULL ) {
-				g_TestRun.frame = atoi(token);
-			}
-		}
-		else if( stricmp(token, "-numimages") == 0 ) {
-			token = argv[i++];
-			if( token != NULL ) {
-				g_TestRun.numimages = atoi(token);
-			}
-		}
-		else if( stricmp(token, "-jpg") == 0 ) {
-			g_TestRun.jpgcapture = 1;
-		}
-		else if( stricmp(token, "-gs") == 0 ) {
-			token = argv[i++];
-			g_TestRun.pgsdll = token;
-		}
-		else if( stricmp(token, "-cdvd") == 0 ) {
-			token = argv[i++];
-			g_TestRun.pcdvddll = token;
+	else if( stricmp(token, "-loadgs") == 0 ) {
+		g_pRunGSState = argv[i++];
+	}
+	else if( stricmp(token, "-pad") == 0 ) {
+		token = argv[i++];
+		printf("-pad ignored\n");
+	}
+#ifdef PCSX2_DEVBUILD
+	else if( stricmp(token, "-numimages") == 0 ) {
+		token = argv[i++];
+		if( token != NULL ) {
+			g_TestRun.numimages = atoi(token);
 		}
-		else if( stricmp(token, "-spu") == 0 ) {
-			token = argv[i++];
-			g_TestRun.pspudll = token;
+	}
+	else if( stricmp(token, "-gs") == 0 ) {
+		token = argv[i++];
+		g_TestRun.pgsdll = token;
+	}
+	else if( stricmp(token, "-cdvd") == 0 ) {
+		token = argv[i++];
+		g_TestRun.pcdvddll = token;
+	}
+	else if( stricmp(token, "-spu") == 0 ) {
+		token = argv[i++];
+		g_TestRun.pspudll = token;
+	}
+	else if( stricmp(token, "-loadgs") == 0 ) {
+		token = argv[i++];
+		g_pRunGSState = token;
+	}
+	else if( stricmp(token, "-image") == 0 ) {
+		g_TestRun.pimagename = argv[i++];
+	}
+	else if( stricmp(token, "-log") == 0 ) {
+		g_TestRun.plogname = argv[i++];
+	}
+	else if( stricmp(token, "-logopt") == 0 ) {
+		token = argv[i++];
+		if( token != NULL ) {
+			if( token[0] == '0' && token[1] == 'x' ) token += 2;
+			sscanf(token, "%x", &varLog);
 		}
-		else if( stricmp(token, "-loadgs") == 0 ) {
-			token = argv[i++];
-			g_pRunGSState = token;
+	}
+	else if( stricmp(token, "-frame") == 0 ) {
+		token = argv[i++];
+		if( token != NULL ) {
+			g_TestRun.frame = atoi(token);
 		}
+	}
+	else if( stricmp(token, "-jpg") == 0 ) {
+		g_TestRun.jpgcapture = 1;
+	}
         else if( stricmp(token, "-test") == 0 ) {
             g_TestRun.enabled = 1;
         }
-		else {
+#endif
+	else {
             file = token;
             printf("opening file %s\n", file);
 		}
diff -BuNr orig/pcsx2/x86/iVUzerorec.cpp src/pcsx2/x86/iVUzerorec.cpp
--- orig/pcsx2/x86/iVUzerorec.cpp	2007-04-01 22:26:46.000000000 -0400
+++ src/pcsx2/x86/iVUzerorec.cpp	2007-08-14 05:37:33.000000000 -0400
@@ -60,7 +60,9 @@
 #include "iVUzerorec.h"
 
 // temporary externs
+#ifdef PCSX2_DEVBUILD
 extern u32 vudump;
+#endif
 extern void iDumpVU0Registers();
 extern void iDumpVU1Registers();
 
@@ -2513,6 +2515,7 @@
 //    VU1.VF[7].F[2] = vuDouble(VU1.VF[7].UL[2]);
 //    VU1.VF[7].F[3] = vuDouble(VU1.VF[7].UL[3]);
 
+#ifdef PCSX2_DEVBUILD
     if( ((vudump&8) && g_curdebugvu) || ((vudump&0x80) && !g_curdebugvu) ) { //&& g_vu1lastrec != g_vu1last ) {
 
 		if( skipparent != g_vu1lastrec ) {
@@ -2534,6 +2537,7 @@
 
 		g_vu1lastrec = s_svulast;
 	}
+#endif
 
 #if defined(_MSC_VER) && !defined(__x86_64__)
 	__asm {
diff -BuNr orig/pcsx2/x86/ix86-32/aR5900-32.S src/pcsx2/x86/ix86-32/aR5900-32.S
--- orig/pcsx2/x86/ix86-32/aR5900-32.S	2007-04-01 22:26:41.000000000 -0400
+++ src/pcsx2/x86/ix86-32/aR5900-32.S	2007-08-14 05:40:07.000000000 -0400
@@ -185,8 +185,10 @@
 	push %ecx
 
 	rdtsc
+#ifdef PCSX2_DEVBUILD
 	mov dword ptr [lbase], %eax
 	mov dword ptr [lbase + 4], %edx
+#endif
 
 	pop %ecx
 	pop %ebx
@@ -203,8 +205,10 @@
 
 	rdtsc
 
+#ifdef PCSX2_DEVBUILD
 	sub %eax, dword ptr [lbase]
 	sbb %edx, dword ptr [lbase + 4]
+#endif
 	mov %ecx, s_pCurBlock_ltime
 	add %eax, dword ptr [%ecx]
 	adc %edx, dword ptr [%ecx + 4]
