--- configure.ac.old	2007-08-14 03:32:44.000000000 -0400
+++ configure.ac	2007-08-14 03:33:24.000000000 -0400
@@ -108,7 +108,12 @@
 AC_CHECK_LIB(stdc++,main,[LIBS="$LIBS -lstdc++"])
 AC_CHECK_LIB(z,main,[LIBS="$LIBS -lz"])
 
-AC_CHECK_HEADER([libintl.h], [AC_DEFINE(ENABLE_NLS,1,[__x86_64__])])
+AC_ARG_ENABLE(nls, AC_HELP_STRING([--disable-nls], [Disable nls support]),
+ nls=$enableval,nls=yes)
+if test "x$nls" == xyes
+then
+ AC_CHECK_HEADER([libintl.h], [AC_DEFINE(ENABLE_NLS,1,[__x86_64__])])
+fi
 
 AC_OUTPUT([
  Makefile
@@ -132,3 +137,4 @@
 echo "  Dev build?	           $devbuild"
 echo "  Recompilers enabled?   $recbuild"
 echo "  Virtual memory build?  $vmbuild"
+echo "  nls support?	       $nls"
