diff -ur pfscalibration-1.3.orig/src/jpeg2hdrgen pfscalibration-1.3/src/jpeg2hdrgen
--- pfscalibration-1.3.orig/src/jpeg2hdrgen	2006-03-16 09:22:46.000000000 -0300
+++ pfscalibration-1.3/src/jpeg2hdrgen	2006-07-23 22:25:55.000000000 -0300
@@ -23,6 +23,8 @@
 
 LC_NUMERIC=POSIX                # ensure dot is used as the decimal separator
 export LC_NUMERIC
+LC_ALL=POSIX                # ensure dot is used as the decimal separator
+export LC_ALL
 
 JHEAD="jhead"                   # program for extracting exif info from jpegs
 
Seulement dans pfscalibration-1.3/src: jpeg2hdrgen~
