diff -ur madwifi-0.9.3.3.orig/ath_hal/ah_target.inc madwifi-0.9.3.3/ath_hal/ah_target.inc
--- madwifi-0.9.3.3.orig/ath_hal/ah_target.inc	2006-11-18 00:34:30.000000000 +0100
+++ madwifi-0.9.3.3/ath_hal/ah_target.inc	2008-01-25 20:34:35.000000000 +0100
@@ -37,5 +37,7 @@
 TARGET-$(CONFIG_CPU_SA110) = xscale-$(ENDIAN)-elf
 TARGET-$(CONFIG_CPU_SA1100) = xscale-$(ENDIAN)-elf
 TARGET-$(CONFIG_CPU_XSCALE) = xscale-$(ENDIAN)-elf
+TARGET-$(CONFIG_X86_64) = x86_64-elf
+TARGET-$(CONFIG_X86_32) = i386-elf
 
 TARGET = $(if $(TARGET-y),$(TARGET-y),$(ARCH)-elf)
diff -ur madwifi-0.9.3.3.orig/scripts/get_arch.mk madwifi-0.9.3.3/scripts/get_arch.mk
--- madwifi-0.9.3.3.orig/scripts/get_arch.mk	2006-11-15 08:24:13.000000000 +0100
+++ madwifi-0.9.3.3/scripts/get_arch.mk	2008-01-25 20:32:23.000000000 +0100
@@ -9,7 +9,7 @@
 ARCH-$(CONFIG_CRIS) = cris
 ARCH-$(CONFIG_FRV) = frv
 ARCH-$(CONFIG_H8300) = h8300
-ARCH-$(CONFIG_X86) = i386
+ARCH-$(CONFIG_X86) = x86
 ARCH-$(CONFIG_IA64) = ia64
 ARCH-$(CONFIG_M32R) = m32r
 ARCH-$(CONFIG_M68K) = m68k
@@ -29,7 +29,7 @@
 ARCH-$(CONFIG_SPARC64) = sparc64
 ARCH-$(CONFIG_UML) = um
 ARCH-$(CONFIG_V850) = v850
-ARCH-$(CONFIG_X86_64) = x86_64
+ARCH-$(CONFIG_X86_64) = x86
 ARCH-$(CONFIG_XTENSA) = xtensa
 
 ifeq (,$(ARCH-y))
