--- wxWidgets/src/gtk/notebook.cpp	2006/03/06 02:20:17	1.120.2.2
+++ wxWidgets/src/gtk/notebook.cpp	2006/11/14 16:23:20	1.120.2.3
@@ -599,7 +599,6 @@
 
     gtk_widget_ref( client->m_widget );
     gtk_widget_unrealize( client->m_widget );
-    gtk_widget_unparent( client->m_widget );
 
     // gtk_notebook_remove_page() sends "switch_page" signal with some strange
     // new page index (when deleting selected page 0, new page is 1 although,
