--- kmobiletools/CMakeLists.txt.org	2008-12-07 16:19:29.000000000 +0100
+++ kmobiletools/CMakeLists.txt	2008-12-08 19:09:55.000000000 +0100
@@ -1,5 +1,54 @@
+INCLUDE("/var/lib/kde/libkdepim:4.1")
+INCLUDE("/var/lib/kde/libkleo:4.1")
 project(kmobiletools)
 
+option(KDE4_KDEPIM_NEW_DISTRLISTS "Whether to use new distribution lists, which stores them like normal contacts; useful for Kolab")
+if(KDE4_KDEPIM_NEW_DISTRLISTS)
+  add_definitions (-DKDEPIM_NEW_DISTRLISTS)
+endif(KDE4_KDEPIM_NEW_DISTRLISTS)
+
+add_definitions ( -DTest1Area=5121 ) # jstaniek
+
+set(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake/modules)
+
+# search packages used by KDE
+find_package (KDE4 REQUIRED)
+include(KDE4Defaults)
+include(MacroLibrary)
+include(MacroOptionalAddSubdirectory)
+include(CheckIncludeFiles)
+include(FindKdepimLibs)
+include_directories (${CMAKE_SOURCE_DIR} ${CMAKE_BINARY_DIR} ${KDEPIMLIBS_INCLUDE_DIR} ${KDE4_INCLUDES} ${QT_QTDBUS_INCLUDE_DIR} ${CMAKE_SOURCE_DIR}/libkdepim/)
+
+add_definitions (${QT_DEFINITIONS} ${KDE4_DEFINITIONS} -DHAVE_CONFIG_H=1)
+
+set(KDE4_ICON_DIR ${KDE4_INSTALL_DIR}/share/icons)
+
+# this macro exists to work around a stupid Mac OS X linker bug
+# where it can't handle the same library being referenced multiple
+# times on the linker line
+macro (KDEPIM4_LINK_UNIQUE_LIBRARIES)
+  set(LINK_ARG_LIST)  
+  foreach(arg ${ARGN})
+    set(LINK_ARG_FOUND FALSE)
+    foreach(existing ${LINK_ARG_LIST})
+      if(${arg} STREQUAL ${existing})
+        set(LINK_ARG_FOUND)
+      endif(${arg} STREQUAL ${existing})
+    endforeach(existing)
+
+    if(NOT ${LINK_ARG_FOUND})
+      list(APPEND LINK_ARG_LIST ${arg})
+    endif(NOT ${LINK_ARG_FOUND})
+  endforeach(arg)
+
+  target_link_libraries(${LINK_ARG_LIST})
+endmacro (KDEPIM4_LINK_UNIQUE_LIBRARIES)
+
+
+
+
+
 include_directories(${CMAKE_CURRENT_BINARY_DIR} ${CMAKE_SOURCE_DIR}/akonadi )
 
 
