--- Miro-0.9.8.orig/platform/gtk-x11/setup.py	2007-07-17 15:32:42.000000000 -0400
+++ Miro-0.9.8/platform/gtk-x11/setup.py	2007-07-19 19:59:12.000000000 -0400
@@ -207,7 +207,7 @@
 if mozIncludeBase is None:
     raise ValueError("Can't find mozilla include base directory")
 for subdir in ['dom', 'gfx', 'widget', 'commandhandler', 'uriloader',
-            'webbrwsr']:
+            'webbrwsr', 'necko']:
     path = os.path.join(mozIncludeBase, subdir)
     mozilla_browser_options['include_dirs'].append(path)
 
@@ -243,7 +243,7 @@
 
 #### Xine Extension ####
 xine_options = parsePkgConfig('pkg-config', 
-        'libxine pygtk-2.0 gtk+-2.0 glib-2.0 gthread-2.0')
+        'libxine pygtk-2.0 gtk+-2.0 glib-2.0 gthread-2.0 x11')
 xine_ext = Extension('miro.xine', [
         os.path.join(xine_dir, 'xine.pyx'),
         os.path.join(xine_dir, 'xine_impl.c'),
