diff -r -u -p kdebase-3.5.6.original/kicker/libkicker/panelbutton.cpp kdebase-3.5.6.patched/kicker/libkicker/panelbutton.cpp
--- kdebase-3.5.6.original/kicker/libkicker/panelbutton.cpp	2006-07-22 10:15:18.000000000 +0200
+++ kdebase-3.5.6.patched/kicker/libkicker/panelbutton.cpp	2007-02-09 21:43:06.000000000 +0100
@@ -29,6 +29,7 @@ CONNECTION WITH THE SOFTWARE OR THE USE 
 #include <qstyle.h>
 #include <qstylesheet.h>
 #include <qtooltip.h>
+#include <qpixmap.h>
 
 #include <kapplication.h>
 #include <kconfig.h>
@@ -39,6 +40,7 @@ CONNECTION WITH THE SOFTWARE OR THE USE 
 #include <kglobalsettings.h>
 #include <kiconloader.h>
 #include <kicontheme.h>
+#include <kiconeffect.h>
 #include <kipc.h>
 #include <kstandarddirs.h>
 #include <klocale.h>
@@ -487,6 +489,9 @@ void PanelButton::mouseReleaseEvent(QMou
     if (e->button() == LeftButton)
     {
         m_isLeftMouseButtonDown = false;
+	
+	QPixmap pix = labelIcon();
+	KIconEffect::visualActivate(this, this->geometry(), &pix);
     }
     QButton::mouseReleaseEvent(e);
 }
