diff -Nurp suse_kickoff_qstyle-nosvn/kicker/kicker/buttons/knewbutton.cpp suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/buttons/knewbutton.cpp
--- suse_kickoff_qstyle-nosvn/kicker/kicker/buttons/knewbutton.cpp     2007-01-28 13:43:47.000000000 +0100
+++ suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/buttons/knewbutton.cpp       2007-01-28 13:44:11.000000000 +0100
@@ -65,7 +65,7 @@ KNewButton::KNewButton( QWidget* parent 
 
     setIconAlignment((Qt::AlignmentFlags)(AlignTop|AlignRight));
     setAcceptDrops(true);
-    setIcon("kmenu-suse");
+    setIcon("kmenu-kickoff");
     setDrawArrow(false);
 
     m_movie = new QMovie(locate("appdata", "pics/kmenu_basic.mng"));
@@ -245,12 +245,12 @@ void KNewButton::drawEye()
             m_iconSize = pixmap.size();
             sf = buttonScaleFactor(size());
 
-            QPainter p(&pixmap);
+/*            QPainter p(&pixmap);
             p.setPen(white);
             p.setBrush(white);
             //      p.setPen(QColor(110,185,55));
             p.drawRect(eye_x+dx, eye_y+dy, 2, 2);
-            p.end();
+            p.end(); */
         }
 
         QWMatrix matrix;
diff -Nurp suse_kickoff_qstyle-nosvn/kicker/kicker/ui/k_new_mnu.cpp suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/k_new_mnu.cpp
--- suse_kickoff_qstyle-nosvn/kicker/kicker/ui/k_new_mnu.cpp   2007-01-27 18:46:31.000000000 +0100
+++ suse_kickoff_qstyle-nosvn-slintegration/kicker/kicker/ui/k_new_mnu.cpp     2007-01-28 13:38:34.000000000 +0100
@@ -1197,7 +1197,9 @@ void KMenu::initialize()
      */
     KConfig ksmserver("ksmserverrc", false, false);
     ksmserver.setGroup("General");
-    connect( m_lock, SIGNAL(clicked()), SLOT(slotOpenSUSE()));
+    //connect( m_lock, SIGNAL(clicked()), SLOT(slotLock()));
+    m_lock->hide();
+
     m_tabBar->setTabEnabled(LeaveTab, kapp->authorize("logout"));
 
     // load search field history
@@ -1331,11 +1331,11 @@
 
     m_systemView->insertSeparator( nId++, i18n("Applications"), index++);
 
-    KService::Ptr p = KService::serviceByStorageId("/usr/share/applications/YaST.desktop");
-    m_systemView->insertMenuItem(p, nId++, index++);
+    m_systemView->insertItem( "kcontrol", i18n( "System Settings" ), i18n( "Control Center" ), 
+    				"/usr/kde/3.5/share/applications/kde/KControl.desktop", nId++, index++ );
 
-    m_systemView->insertItem( "info", i18n( "System Information" ),
-                              "sysinfo:/",  "sysinfo:/", nId++, index++ );
+    m_systemView->insertItem( "kcmprocessor", i18n( "System Information" ),
+                              "Information Center",  "/usr/kde/3.5/share/applications/kde/kinfocenter.desktop", nId++, index++ );
 
     m_systemView->insertSeparator( nId++, i18n("System Folders"), index++ );

--- kicker-3.5.6/kicker/kicker/ui/appletop_mnu.cpp.orig	2007-02-12 12:20:59.000000000 -0330
+++ kicker-3.5.6/kicker/kicker/ui/appletop_mnu.cpp	2007-02-12 12:23:52.000000000 -0330
@@ -169,9 +169,9 @@
         }
 
         if (KickerSettings::legacyKMenu()) 
-          insertItem(SmallIcon("suse"), i18n("Switch to SUSE Menu Style"), this, SLOT(toggleLegacy()));
+          insertItem(SmallIcon("about_kde"), i18n("Switch to New Kickoff"), this, SLOT(toggleLegacy()));
         else
-          insertItem(SmallIcon("about_kde"), i18n("Switch to KDE Menu Style"), this, SLOT(toggleLegacy()));
+          insertItem(SmallIcon("about_kde"), i18n("Switch to Legacy Menu"), this, SLOT(toggleLegacy()));
     }
 
     if ((actions & PanelAppletOpMenu::KMenuEditor) && kapp->authorizeKAction("menuedit"))
--- kicker-3.5.6/kicker/libkicker/panelbutton.cpp.orig	2007-02-09 15:04:39.000000000 -0330
+++ kicker-3.5.6/kicker/libkicker/panelbutton.cpp	2007-02-09 15:05:23.000000000 -0330
@@ -809,7 +809,7 @@
     QString nm = m_iconName;
     KIcon::States defaultState = isEnabled() ? KIcon::DefaultState :
                                                KIcon::DisabledState;
-    if (nm=="kmenu-suse")
+    if (nm=="kmenu-kickoff")
     {
         QString pth = locate( "appdata", "pics/kmenu_basic.mng" );
         if (!pth.isEmpty())
--- kicker-3.5.6/kicker/libkicker/kickerSettings.kcfg.orig	2007-02-12 12:52:15.000000000 -0330
+++ kicker-3.5.6/kicker/libkicker/kickerSettings.kcfg	2007-02-12 12:55:23.000000000 -0330
@@ -104,12 +104,12 @@
    </entry>
 
 <entry name="OpenOnHover" type="Bool" >
-      <label>When this option is enabled, the SUSE Menu does open on mouse hover.</label>
-      <default>true</default>
+      <label>When this option is enabled, the Kickoff Menu does open on mouse hover.</label>
+      <default>false</default>
    </entry>
 
 <entry name="ScrollFlipView" type="Bool" >
-      <label>When this option is enabled, the SUSE Menu application view switching will scroll.</label>
+      <label>When this option is enabled, the Kickoff Menu application view switching will scroll.</label>
       <default>true</default>
    </entry>
 
@@ -148,7 +148,7 @@
 
 <entry name="KickoffSwitchTabsOnHover" type="Bool" >
       <label>When this option is enabled, the tabs in the Kickoff menu will switch without the need to click</label>
-      <default>true</default>
+      <default>false</default>
    </entry>
 
 
