--- Makefile.am.orig	2007-05-29 18:26:31.000000000 +0200
+++ Makefile.am	2007-05-29 18:26:58.000000000 +0200
@@ -29,12 +29,12 @@
 distcheck-hook: libpurple/plugins/perl/common/Purple.pm pidgin/plugins/perl/common/Pidgin.pm
 #	cp libpurple/plugins/perl/common/Gaim.pm $(distdir)/libpurple/plugins/perl/common
 
+if ENABLE_GTK
 appsdir = $(datadir)/applications
 apps_in_files = pidgin.desktop.in
 apps_DATA = $(apps_in_files:.desktop.in=.desktop)
 @INTLTOOL_DESKTOP_RULE@
 
-if ENABLE_GTK
 GTK_DIR=pidgin
 endif