diff -Naur a/mosquitto-1.4/src/CMakeLists.txt b/mosquitto-1.4/src/CMakeLists.txt
--- a/mosquitto-1.4/src/CMakeLists.txt	2015-02-18 02:44:09.000000000 +0100
+++ b/mosquitto-1.4/src/CMakeLists.txt	2015-03-25 16:21:52.330872171 +0100
@@ -113,7 +113,7 @@
 	if (APPLE)
 		set_target_properties(mosquitto PROPERTIES LINK_FLAGS "-Wl,-exported_symbols_list -Wl,${mosquitto_SOURCE_DIR}/src/linker-macosx.syms")
 	else (APPLE)
-		set_target_properties(mosquitto PROPERTIES LINK_FLAGS "-Wl,-exported_symbols_list -Wl,${mosquitto_SOURCE_DIR}/src/linker.syms")
+		#set_target_properties(mosquitto PROPERTIES LINK_FLAGS "-Wl,-exported_symbols_list -Wl,${mosquitto_SOURCE_DIR}/src/linker.syms")
 	endif (APPLE)
 endif (UNIX)
 
