diff -Naur k3b-2.0_pre20080904/plugins/decoder/ffmpeg.org/k3bffmpegdecoder.cpp k3b-2.0_pre20080904/plugins/decoder/ffmpeg/k3bffmpegdecoder.cpp
--- k3b-2.0_pre20080904/plugins/decoder/ffmpeg.org/k3bffmpegdecoder.cpp	2008-10-08 14:07:51.000000000 +0200
+++ k3b-2.0_pre20080904/plugins/decoder/ffmpeg/k3bffmpegdecoder.cpp	2008-10-08 14:12:49.000000000 +0200
@@ -21,8 +21,8 @@
 #include <kdebug.h>
 
 extern "C" {
-#include <ffmpeg/avcodec.h>
-//#include <libavcodec/avcodec.h>
+//#include <ffmpeg/avcodec.h>
+#include <libavcodec/avcodec.h>
 }
 
 #include <math.h>
diff -Naur k3b-2.0_pre20080904/plugins/decoder/ffmpeg.org/k3bffmpegwrapper.cpp k3b-2.0_pre20080904/plugins/decoder/ffmpeg/k3bffmpegwrapper.cpp
--- k3b-2.0_pre20080904/plugins/decoder/ffmpeg.org/k3bffmpegwrapper.cpp	2008-10-08 14:10:21.000000000 +0200
+++ k3b-2.0_pre20080904/plugins/decoder/ffmpeg/k3bffmpegwrapper.cpp	2008-10-08 14:12:49.000000000 +0200
@@ -17,10 +17,10 @@
 #include <config-k3b.h>
 
 extern "C" {
-#include <ffmpeg/avcodec.h>
-#include <ffmpeg/avformat.h>
-// #include <libavcodec/avcodec.h>
-// #include <libavformat/avformat.h>
+//#include <ffmpeg/avcodec.h>
+//#include <ffmpeg/avformat.h>
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
 }
 
 #include <string.h>
