diff -ur coremoid-0.3.1.org/coremoidapplet.cpp coremoid-0.3.1/coremoidapplet.cpp
--- coremoid-0.3.1.org/coremoidapplet.cpp	2008-08-20 11:35:35.000000000 +0200
+++ coremoid-0.3.1/coremoidapplet.cpp	2008-08-20 11:36:44.000000000 +0200
@@ -103,8 +103,8 @@
 	m_boxAlpha = 128;
 	m_boxHoverAlpha = 192;
 	m_boxColor.setAlpha(m_boxAlpha);
-	m_textColor = Plasma::Theme::self()->textColor();
-	m_boxColor = Plasma::Theme::self()->backgroundColor();
+	m_textColor = Plasma::Theme::defaultTheme()->color(Plasma::Theme::TextColor);
+	m_boxColor = Plasma::Theme::defaultTheme()->color(Plasma::Theme::BackgroundColor);
 
 	setAcceptsHoverEvents(true);
 	setHasConfigurationInterface(true);
