diff -Naur a/ozwcp.cpp b/ozwcp.cpp
--- a/ozwcp.cpp	2014-04-24 13:52:20.732642960 +0200
+++ b/ozwcp.cpp	2014-04-24 13:54:00.205649898 +0200
@@ -774,7 +774,7 @@
   for (i = 0; i < MAX_NODES; i++)
     nodes[i] = NULL;
 
-  Options::Create("./config/", "", "--SaveConfiguration=true --DumpTriggerLevel=0");
+  Options::Create("~/.config/openzwave-control-panel/", "", "--SaveConfiguration=true --DumpTriggerLevel=0");
   Options::Get()->Lock();
 
   Manager::Create();
